ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto

```diff
@@ -95,6 +127,38 @@ message Commands {
   // to read contract data on behalf of each of the given parties.
   // Optional
   repeated string read_as = 13;
+
+  // A unique identifier to distinguish completions for different submissions with the same change ID.
+  // Typically a random UUID. Applications are expected to use a different UUID for each retry of a submission
+  // with the same change ID.
+  //
+  // If omitted, the participant or Daml ledger may set a UUID of their choice.
+  // Optional
+  //
+  // TODO(v2) Do we want to enforce the format of a UUID here?
+  string submission_id = 17;
+
+  // Applications can rank their submissions with the same change ID.
+  // This allows the participants and Daml ledger to detect when one submission has been overtaken by a later resubmission.
+  // The participant and Daml ledger provide the following rank guarantee:
+  // Let there be two submissions for the same change ID with submission ranks `rank1 <= rank2` and submission IDs `sid1` and `sid2`.
+  // Suppose that the submission `sid2` has completed with a definite answer on the completion stream at offset `off2`.
+  // If there is no completion for the submission `sid1` before `off2`, then the submission `sid1` will result at most
+  // in a cancellation (it might be that the submission was lost and no result will ever reach the application).
```

Comment:
'cancellation' is not defined in the context of these Ledger API gRPC services. Consider referring to the `ALREADY_EXISTS` error.

Reply:
Rephrased into `definite answer`. `ALREADY_EXISTS` is rather orthogonal to whether the completion is a definite answer.